src/fidder/erase/sparse_local_mean.py

```diff
@@ -59,3 +60,71 @@ def estimate_local_mean(
     x = np.arange(image.shape[-1])
     local_mean = background_model(y, x, grid=True)
     return torch.tensor(local_mean, dtype=input_dtype)
+
+
+def estimate_local_mean_3d(
+    volume: torch.Tensor,
+    mask: Optional[torch.Tensor] = None,
+    resolution: Tuple[int, int, int] = (5, 5, 5),
+    n_samples_for_fit: int = 20000,
+):
+    """Estimate local mean of an image with a bivariate cubic spline.
+
+    A mask can be provided to
+
+    Parameters
+    ----------
+    image: torch.Tensor
+        `(h, w)` array containing image data.
+    mask: Optional[torch.Tensor]
+        `(h, w)` array containing a binary mask specifying foreground
+        and background pixels for the estimation.
+    resolution: Tuple[int, int]
+        Resolution of the local mean estimate in each dimension.
+    n_samples_for_fit: int
+        Number of samples taken from foreground pixels for background mean estimation.
+        The number of background pixels will be used if this number is greater than the
+        number of background pixels.
+
+    Returns
+    -------
+    local_mean: torch.Tensor
+        `(h, w)` array containing a local estimate of the local mean.
+    """
+    input_dtype = volume.dtype
+    volume = volume.numpy()
+    mask = np.ones_like(volume) if mask is None else mask.numpy()
+
+    # get a random set of foreground pixels for the background fit
+    foreground_sample_idx = np.argwhere(mask == 1)
+
+    n_samples_for_fit = min(n_samples_for_fit, len(foreground_sample_idx))
+    selection = np.random.choice(
+        foreground_sample_idx.shape[0], size=n_samples_for_fit, replace=False
+    )
+    foreground_sample_idx = foreground_sample_idx[selection]
+    z, y, x = foreground_sample_idx[:, 0], foreground_sample_idx[:, 1], foreground_sample_idx[:, 2]
+
+    w = torch.as_tensor(volume[(z, y, x)])
+
+    grid = CubicBSplineGrid3d(resolution=resolution)
+    optimiser = torch.optim.Adam(grid.parameters(), lr=0.01)
+
+    for i in range(500):
+        # what does the model predict for our observations?
+        prediction = grid(foreground_sample_idx).squeeze()
```

Comment:
this looks like it's about right except when sampling the model coordinates should be rescaled to [0, 1] where 0 is the center of the first element in the dimension and 1 is the center of the last element in the dimension

Reply:
Mhmh... I see, I'll look into it :)

Reply:
@alisterburt does this make sense now? Though unfortunately it doens't look better  @alessiodacapito :/ I think the bulk of the work here is being done by the noise generation, not the local mean...